content/blog/2019-10-30-loop-reduction/index.md

```diff
@@ -0,0 +1,375 @@
+
+   Assume the cost of `x = y*z` is `c1` and that of `x = y+z` is `c2`. We reduce the computation in the program by `n*(c1-c2) - c1`. To benchmark this, we timed our optimized version as in Loop Invariant Code Motion. The results are
+
+   ```
+   hyperfine --warmup 1 'brili < orig_strengthreduction1'
+   Time (mean ± σ):     192.1 ms ±   7.2 ms
+   
+   hyperfine --warmup 1 'brili < opt_strengthreduction1'
+   Time (mean ± σ):     173.2 ms ±   10.3 ms
+   ```
+
+   Similarly, `strengthreduction3.ts` has division to be reduced.
+
+   ```
+   hyperfine --warmup 1 'brili < orig_strengthreduction2'
+   Time (mean ± σ):     61.1 ms ±   2.4 ms
+   
+   hyperfine --warmup 1 'brili < opt_strengthreduction2'
+   Time (mean ± σ):     56.3 ms ±   5.9 ms
+   ```
+
+   ### Both Optimizations Examples
+
+   To combine both loop optimizations, namely loop invariant code motion and strength reduction, we run tests `both1` and `both2` following the same evaluation method. The results are as follows.
+
+   ```
+   hyperfine --warmup 1 'brili < orig_both1'
+   Time (mean ± σ):     655.0 ms ±   8.6 ms
+   
+   hyperfine --warmup 1 'brili < opt_both1'
+   Time (mean ± σ):     484.1 ms ±   13.7 ms
+   ```
+
+   ```
+   hyperfine --warmup 1 'brili < orig_both2'
+   Time (mean ± σ):     66.2 ms ±   3.3 ms
+   
+   hyperfine --warmup 1 'brili < opt_both2'
+   Time (mean ± σ):     49.8 ms ±   5.9 ms
+   ```
+
+   ### Final Results
+
+   | Filename                | Unoptimized runtime | Optimized runtime | Speedup |
+   | ----------------------- | ------------------- | ----------------- | ------- |
+   | `codemotion1.ts`        | 216.8 ms            | 163.8 ms          | 1.32x   |
+   | `codemotion2.ts`        | 908.6 ms            | 665.4 ms          | 1.36x   |
+   | `strengthreduction1.ts` | 192.1 ms            | 173.2 ms          | 1.11x   |
+   | `strengthreduction2.ts` | 61.1 ms             | 56.3 ms           | 1.09x   |
+   | `both1.ts`              | 655.0 ms            | 484.1 ms          | 1.35x   |
+   | `both2.ts`              | 66.2 ms             | 49.8 ms           | 1.33x   |
+
+   Based on the evaluation results, we find out that both loop optimization techniques can provide some speedup. Loop invariant code motion can provide up to 1.36x speedup, while strength reduction can only have about 10% performance improvement. That is due to the fact that multiplication/division does not take significantly more cycles than addition/substraction on modern machines and there are many redundant `id` operations when generating `bril` programs from `ts`. Thus, only limited performance improvement can be obtained We expect that more speedup can be obtained if exponential operation is optimized. Also, note that there are nested loops in `codemotion2.ts` and `both1.ts`, thus their runtime are much longer than the rest.
```

Comment:
This is a good, detailed evaluation! Nice work. You might have gotten a little more detail about the performance by tracing the counts of instruction types—for example, showing that you do decrease the number of multiplication instructions executed could be cool, even if the number add instructions goes up.